### PR TITLE
add print.mmrm_tmb method

### LINKED ISSUE #31: Add `print` method for `mmrm_tmb`
This is nice for handling the results and will make testing of the package easier.

To do:

- [x] see how other packages are doing it, e.g.:
	- [x] `glmmTMB`: glmmTMB uses `print.glmmTMB` to present their objects, with the typical model output that one would expect. Very similar to what is spec'd below. 
	- [x] `lme4`: for lmList4 they use `show` to print out a model summary (coded within the function body), for merMod they first spec a S3 `print` method, and then call that in the S4 `show` method body. The difference seems to be that lmList4 inherits the print method from `lme` objects from the `nmle` package. 
- [x] do a nice first version (does not need to be perfect)
	- [x] in https://github.com/openpharma/mmrm/blob/main/R/tmb-methods.R
  - [x] include formula, data, covariance structure (can reuse functions from printing the mmrm object)
  - [x] plus include REML vs ML used, optimizer used, convergence yes/no
  - [x] iterations counts
  - [x] see other tmb methods for how to document it
	- [x] add test in corresponding test file https://github.com/openpharma/mmrm/blob/main/tests/testthat/test-tmb-methods.R

### PR DESCRIPTION
**Pull request by BrianLang**
closes #31 

added `print.mmrm_tmb` method.
updated `h_mmrm_tmb_fit` to return call which allowed us to use `h_print_call`.
updated testing to cope with `h_mmrm_tmb_fit` update.
added summary.mmrm_tmb method to cope with differences btween mmrm and mmrm_tmb objects.

Still need testing, but want to check that we are on the right track first. 

---

Made a first draft following what has been done for `mmrm` objects. Needed to update the `mmrm_tmb` object definition in `h_mmrm_tmb_fit` to allow for use of `h_print_call`. 

```
> mmrm_tmb_obj <- get_mmrm_tmb()
> mmrm_tmb_obj
mmrm_tmb fit

Formula:     FEV1 ~ RACE + us(AVISIT | USUBJID)
Data:        data (used 537 observations from 197 subjects with maximum 4 timepoints)
Covariance:  unstructured (1 variance parameters)

Model selection criteria:
     AIC      BIC   logLik deviance 
  3662.4   3695.2  -1821.2   3642.4 

Coefficients:
                  (Intercept) RACEBlack or African American                     RACEWhite 
                    41.227284                      0.800180                      5.879103 

Model Inference Optimization:
Optimizer: BFGS
Method: REML
Converged with code  0  and message:  relative convergence (4)
```

```
> mmrm_obj <- get_mmrm()
> mmrm_obj
mmrm fit

Formula:     FEV1 ~ RACE + SEX + ARMCD * AVISIT + us(AVISIT | USUBJID)
Data:        fev_data (used 537 observations from 197 subjects with maximum 4 timepoints)
Covariance:  unstructured (1 variance parameters)

Model selection criteria:
     AIC      BIC   logLik deviance 
  3406.4   3439.3  -1693.2   3386.4 

Coefficients:
                  (Intercept) RACEBlack or African American                     RACEWhite 
                  30.77747548                    1.53049977                    5.64356535 
                    SEXFemale                      ARMCDTRT                    AVISITVIS2 
                   0.32606192                    3.77423004                    4.83958845 
                   AVISITVIS3                    AVISITVIS4           ARMCDTRT:AVISITVIS2 
                  10.34211288                   15.05389826                   -0.04192625 
          ARMCDTRT:AVISITVIS3           ARMCDTRT:AVISITVIS4 
                  -0.69368537                    0.62422703 

Model Inference Optimization:
Optimizer: BFGS
Method: REML
Converged with code  0  and message:  convergence: rel_reduction_of_f <= factr*epsmch
```

Have not yet added iteration information since `mmrm` and `mmrm_tmb` objects have different `$optdetails` pertaining to iterations. 

```
> mmrm_tmb_obj <- get_mmrm_tmb()
> mmrm_tmb_obj$opt_details
$convergence
[1] 0

$iterations
[1] 27

$evaluations
function gradient 
      40       28 

$message
[1] "relative convergence (4)"
```

```
> mmrm_obj <- get_mmrm_tmb()
> mmrm_obj$opt_details
$counts
function gradient 
      15       15 

$convergence
[1] 0

$message
[1] "CONVERGENCE: REL_REDUCTION_OF_F <= FACTR*EPSMCH"
```

### CODE REVIEW

File: R/tmb-methods.R [hunk 1]
```diff
@@ -157,3 +157,66 @@ BIC.mmrm_tmb <- function(object, ...) { # nolint
   k <- log(object$tmb_data$n_subjects)
   AIC(object, corrected = FALSE, k = k)
 }
+
+#' @describeIn mmrm_tmb_methods prints the MMRM_TMB object.
+#' @exportS3Method
+#' @keywords internal
+print.mmrm_tmb <- function(x,
+                           ...) {
+  summary_m <- summary(x)
```
Comment (danielinteractive) [inline R/tmb-methods.R:166]: this we don't need then, see my comment below
```suggestion
```

File: R/tmb-methods.R [hunk 2]
```diff
@@ -157,3 +157,66 @@ BIC.mmrm_tmb <- function(object, ...) { # nolint
...
+
+  # cat(paste0("Formula:     ", sub(x = toString(x$formula_parts$formula),
+  #                             pattern = "^(.),\\s(.+),\\s(.+)$",
+  #                             replacement = "\\2 \\1 \\3\n")))
+
+  h_print_call(summary_m$call, summary_m$n_obs, summary_m$n_subjects, summary_m$n_timepoints)
+  h_print_cov(summary_m$cov_type, length(summary_m$n_theta))
+
+  cat("\nModel selection criteria:\n")
+  h_print_aic_list(summary_m$aic_list)
```
Comment (danielinteractive) [inline R/tmb-methods.R:177]: almost too much here, I would just give the obtained deviance

File: R/tmb-methods.R [hunk 3]
```diff
@@ -157,3 +157,66 @@ BIC.mmrm_tmb <- function(object, ...) { # nolint
...
+  print(coef(x))
+
+  cat("\nModel Inference Optimization:\n")
+  cat("Optimizer: ")
+  cat(x$tmb_object$method)
+  cat("\nMethod: ")
+  cat(ifelse(x$reml, "REML", "ML"))
+  cat(ifelse(x$opt_details$convergence == 0, "\nConverged", "\nFailed to converge"))
+  cat(" with code ", x$opt_details$convergence,
+      " and message: ", tolower(x$opt_details$message))
```
Comment (danielinteractive) [inline R/tmb-methods.R:189]: this is good

**Review by danielinteractive — COMMENTED** [review]
thanks Brian! very nice start. Please see my comments inside, let me know if any questions.

File: R/tmb.R [hunk 4]
```diff
@@ -347,8 +347,12 @@ h_mmrm_tmb_fit <- function(tmb_object,
     c("par", "objective")
   )
 
+  fun_call <- match.call()
+  fun_call$formula <- eval(formula_parts$formula)
+
   structure(
     list(
+      call = fun_call,
```
[the remark this answers is not in the document]
Comment (BrianLang) [reply R/tmb.R:359]: This is an interesting thing. 
User-facing functions (`mmrm`, in this case) are the only ones that can consistently return the correct dataset names for match.call(). Everything else is too deep to return anything other than data = "data".
To work around this, I have `mmrm` add the user-supplied dataset name to the dataset as an attribute, which is then accessed by `h_mmrm_tmb_fit` if it exists.

File: R/tmb-methods.R [hunk 5]
```diff
@@ -157,3 +157,66 @@ BIC.mmrm_tmb <- function(object, ...) { # nolint
...
+print.mmrm_tmb <- function(x,
+                           ...) {
+  summary_m <- summary(x)
+  cat(ifelse("mmrm" %in% class(x), "mmrm fit\n\n", "mmrm_tmb fit\n\n"))
+
+  # cat(paste0("Formula:     ", sub(x = toString(x$formula_parts$formula),
+  #                             pattern = "^(.),\\s(.+),\\s(.+)$",
+  #                             replacement = "\\2 \\1 \\3\n")))
+
+  h_print_call(summary_m$call, summary_m$n_obs, summary_m$n_subjects, summary_m$n_timepoints)
```
[the remark this answers is not in the document]
Comment (BrianLang) [reply R/tmb-methods.R]: took approach 1.
getME can be done later

File: R/tmb-methods.R (hunk 3 above)
Comment (BrianLang) [reply R/tmb-methods.R]: thanks :-)

**Comment by BrianLang** [comment]
Updated for feedback

```
> get_mmrm()
mmrm fit

Formula:     FEV1 ~ RACE + SEX + ARMCD * AVISIT + us(AVISIT | USUBJID)
Data:        fev_data (used 537 observations from 197 subjects with maximum 4 timepoints)
Covariance:  unstructured (10 variance parameters)
Model deviance: 3386.45

Coefficients:
                  (Intercept) RACEBlack or African American                     RACEWhite 
                  30.77747548                    1.53049977                    5.64356535 
                    SEXFemale                      ARMCDTRT                    AVISITVIS2 
                   0.32606192                    3.77423004                    4.83958845 
                   AVISITVIS3                    AVISITVIS4           ARMCDTRT:AVISITVIS2 
                  10.34211288                   15.05389826                   -0.04192625 
          ARMCDTRT:AVISITVIS3           ARMCDTRT:AVISITVIS4 
                  -0.69368537                    0.62422703 

Model Inference Optimization:
Optimizer: BFGS
Method: REML
Converged with code  0  and message:  convergence: rel_reduction_of_f <= factr*epsmch
```

```
> get_mmrm_tmb()
mmrm fit

Formula:     FEV1 ~ RACE + us(AVISIT | USUBJID)
Data:        data (used 537 observations from 197 subjects with maximum 4 timepoints)
Covariance:  unstructured (10 variance parameters)
Model deviance: 3642.395

Coefficients:
                  (Intercept) RACEBlack or African American                     RACEWhite 
                    41.227284                      0.800180                      5.879103 

Model Inference Optimization:
Optimizer: BFGS
Method: REML
Converged with code  0  and message:  relative convergence (4)
```

File: R/tmb-methods.R [hunk 6]
```diff
@@ -157,3 +157,41 @@ BIC.mmrm_tmb <- function(object, ...) { # nolint
...
+  n_subjects <- x$tmb_data$n_subjects
+  n_timepoints <- x$tmb_data$n_visits
+  n_obs <- length(x$tmb_data$y_vector)
+  cov_type <- x$formula_parts$corr_type
+  n_theta <- length(x$theta_est)
+
+  h_print_call(x$call, n_obs, n_subjects, n_timepoints)
+  h_print_cov(cov_type, n_theta)
+
+  cat("Model deviance: ")
```
Comment (danielinteractive) [inline R/tmb-methods.R:178]:
```suggestion
  cat("Deviance: ")
```
? just to make it shorter

**Review by danielinteractive — APPROVED** [review]
thanks Brian! ready to merge after last small comments inside resolved. please also put your name and email in the authors at https://github.com/openpharma/mmrm/blob/25a3f3b8417bffaa95247de4377fd6871b939f8c/DESCRIPTION#L8 (after Julia), thanks!

**Comment by danielinteractive** [comment]
@BrianLang you can squash and merge when everything is done, as I approved it already.

**Pull request merged**